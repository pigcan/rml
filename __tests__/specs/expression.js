const { transformExpression } = require('../../src/expression');

describe('transformExpression', () => {
  it('keep this', () => {
    expect(transformExpression('{{this.x}}')).toMatchSnapshot();
  });
  it('support literal', () => {
    expect(transformExpression('x')).toMatchSnapshot();
  });
  it('support simple', () => {
    expect(transformExpression('{{x}}')).toMatchSnapshot();
  });
  it('support forceObject', () => {
    expect(transformExpression('{{x}}', null, { forceObject: 1 })).toMatchSnapshot();
  });
  it('support spread', () => {
    expect(transformExpression('{{ ...x }}')).toMatchSnapshot();
  });
  it('support interpolate', () => {
    expect(transformExpression('1-{{x + y}}-2')).toMatchSnapshot();
  });
  it('support interpolate 2', () => {
    expect(transformExpression('{{x + y}}-2')).toMatchSnapshot();
  });
  it('support ternary', () => {
    expect(transformExpression('{{flag ? true : false}}')).toMatchSnapshot();
  });
  it('support logic', () => {
    expect(transformExpression('{{x>5}}')).toMatchSnapshot();
  });
  it('support string', () => {
    expect(transformExpression('{{"hello"+x}}')).toMatchSnapshot();
  });
  it('support array', () => {
    expect(transformExpression('{{[x,2]}}')).toMatchSnapshot();
  });
  describe('object', () => {
    it('support simple', () => {
      expect(transformExpression('{{x:y}}')).toMatchSnapshot();
    });
    it('support shorthand', () => {
      expect(transformExpression('{{x,y}}')).toMatchSnapshot();
    });
    it('support spread', () => {
      expect(transformExpression('{{...x,y}}')).toMatchSnapshot();
    });
  });
  describe('member expression', () => {
    it('support member function', () => {
      expect(transformExpression('{{x.y.join(",")}}')).toMatchSnapshot();
    });
    it('support simple', () => {
      expect(transformExpression('{{x.a}}')).toMatchSnapshot();
    });
    it('support []', () => {
      expect(transformExpression('{{x[a].b}}')).toMatchSnapshot();
    });
    it('support interpolate', () => {
      expect(transformExpression('1{{x[a].b}}2')).toMatchSnapshot();
    });
  });

  describe('loose data member', () => {
    const args = [null, {
      strictDataMember: false,
    }];
    it('support member function', () => {
      expect(transformExpression('{{x.y.join(",")}}', ...args)).toMatchSnapshot();
    });
    it('support simple', () => {
      expect(transformExpression('{{x}}', ...args)).toMatchSnapshot();
    });
    it('support simple member', () => {
      expect(transformExpression('{{x.a}}', ...args)).toMatchSnapshot();
    });
    it('support expression member', () => {
      expect(transformExpression('{{x.a + x[b]}}', ...args)).toMatchSnapshot();
    });
    it('support [] member', () => {
      expect(transformExpression('{{x[a].b}}', ...args)).toMatchSnapshot();
    });
    it('support interpolate member', () => {
      expect(transformExpression('1{{x[a].b}}2', ...args)).toMatchSnapshot();
    });
  });
});
