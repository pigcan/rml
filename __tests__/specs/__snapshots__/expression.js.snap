// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformExpression keep this 1`] = `"(this.x)"`;

exports[`transformExpression loose data member ignore member computed function 1`] = `"(data.x[data.y.q].join($getLooseDataMember(data.q, \\"a\\")) + $getLooseDataMember(data.z, \\"a\\"))"`;

exports[`transformExpression loose data member ignore member function 1`] = `"(data.x.y.join($getLooseDataMember(data.q, \\"a\\")) + $getLooseDataMember(data.z, \\"a\\"))"`;

exports[`transformExpression loose data member support [] member 1`] = `"($getLooseDataMember(data.x, data.a, \\"b\\"))"`;

exports[`transformExpression loose data member support expression member 1`] = `"($getLooseDataMember(data.x, \\"a\\") + $getLooseDataMember(data.x, data.b))"`;

exports[`transformExpression loose data member support interpolate member 1`] = `"'1' + ($getLooseDataMember(data.x, data.a, \\"b\\")) + '2'"`;

exports[`transformExpression loose data member support simple 1`] = `"(data.x)"`;

exports[`transformExpression loose data member support simple member 1`] = `"($getLooseDataMember(data.x, \\"a\\"))"`;

exports[`transformExpression member expression support [] 1`] = `"(data.x[data.a].b)"`;

exports[`transformExpression member expression support interpolate 1`] = `"'1' + (data.x[data.a].b) + '2'"`;

exports[`transformExpression member expression support member function 1`] = `"(data.x.y.join(\\",\\"))"`;

exports[`transformExpression member expression support simple 1`] = `"(data.x.a)"`;

exports[`transformExpression object support shorthand 1`] = `"(({ x: data.x, y: data.y }))"`;

exports[`transformExpression object support simple 1`] = `"(({ x: data.y }))"`;

exports[`transformExpression object support spread 1`] = `"(({ ...data.x, y: data.y }))"`;

exports[`transformExpression support array 1`] = `"([data.x, 2])"`;

exports[`transformExpression support forceObject 1`] = `"(({ x: data.x }))"`;

exports[`transformExpression support interpolate 1`] = `"'1-' + (data.x + data.y) + '-2'"`;

exports[`transformExpression support interpolate 2 1`] = `"(data.x + data.y) + '-2'"`;

exports[`transformExpression support literal 1`] = `"'x'"`;

exports[`transformExpression support logic 1`] = `"(data.x > 5)"`;

exports[`transformExpression support simple 1`] = `"(data.x)"`;

exports[`transformExpression support spread 1`] = `"(({ ...data.x }))"`;

exports[`transformExpression support string 1`] = `"(\\"hello\\" + data.x)"`;

exports[`transformExpression support ternary 1`] = `"(data.flag ? true : false)"`;
